### ISSUE TITLE
Metrics scraping takes 5 seconds

### ISSUE THREAD

**Issue by karalabe** [sigp/lighthouse#4230]
## Description

I have a mainnet beacon node running Lighthouse. Just a simple new node, nothing fancy, no validators or anything. Enabling metrics and pointing Prometheus to it, it takes about 5 seconds to scrape the metrics. The counterpart execution metrics scraping in Geth takes less than 5ms in comparison.

## Version

Latest docker image, Lighthouse/v4.1.0-693886b

The machine it's running on is a monster server with PCIe4 NVMe, so there's not much room for hardware limitation.

## Present Behaviour

Start Lighthouse with

```
lighthouse bn --network mainnet --execution-endpoint=http://...:8551 --execution-jwt=.../jwtsecret --checkpoint-sync-url=https://mainnet.checkpoint.sigp.io --metrics --metrics-address=0.0.0.0 --metrics-allow-origin=*
```

Metrics scarping takes 5 seconds, everything is in sync and idling.

## Expected Behaviour

It should take milliseconds at worse.

## Steps to resolve

You tell me :)

**Comment by michaelsproul** [comment]
@karalabe How are you scraping the endpoint? When I scrape it with `curl` on my machine it takes no more than 135ms, often less.

I'm using this command (`fish`):

```fish
while true; time curl -s "http://localhost:5054/metrics" > /dev/null; end
```

**Comment by karalabe** [comment]
I'm using Prometheus to scrape and these are the stats from it:

<img width="1553" alt="Screenshot 2023-04-25 at 18 49 10" src="https://user-images.githubusercontent.com/129561/234332787-d84ffe43-af58-4870-b61c-addaf6d8a054.png">

**Comment by karalabe** [comment]
```
while true; do time curl -s "http://localhost:5054/metrics" > /dev/null; done

0.00user 0.00system 0:05.30elapsed 0%CPU (0avgtext+0avgdata 9868maxresident)k
0inputs+0outputs (0major+604minor)pagefaults 0swaps

0.00user 0.00system 0:05.13elapsed 0%CPU (0avgtext+0avgdata 9828maxresident)k
0inputs+0outputs (0major+599minor)pagefaults 0swaps

0.00user 0.00system 0:05.33elapsed 0%CPU (0avgtext+0avgdata 9628maxresident)k
0inputs+0outputs (0major+598minor)pagefaults 0swaps
```

**Comment by paulhauner** [comment]
Interesting. We're not seeing anything like this on our fleet:

<img width="1362" alt="Screenshot 2023-04-26 at 9 36 10 am" src="https://user-images.githubusercontent.com/6660660/234430199-9571c9bf-7d3a-4385-a121-527b2f5320af.png">

The higher times are due to latency between Prometheus and the host (they're on opposite sides of the world). Our timeout is 10s so peaks are probably network timeouts. Here's the local run of the highest value on that graph, indicating that network latency is a primary factor:

```bash
#  Goerli "deep" node

time curl -s http://localhost:5054/metrics

real	0m0.926s
```

The "validator monitor" metrics can be quite intense, although it doesn't seem like you have them enabled @karalabe.

To get an idea of the volume of your metrics I'd be interested in the output of these two commands, please:

```bash
# This returns 4617 for me (Goerli "deep" node).

curl http://localhost:5054/metrics | grep validator_monitor | wc -l
```


```bash
# This returns 10799 for me (Goerli "deep" node).

curl http://localhost:5054/metrics | wc -l
```

**Comment by karalabe** [comment]
I've tried looking at the output but it's small, a couple K lines at max. Also no validator running.

**Comment by karalabe** [comment]
```
# curl http://localhost:5054/metrics | grep validator_monitor | wc -l
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 83120  100 83120    0     0  14081      0  0:00:05  0:00:05 --:--:-- 17681
3
# curl http://localhost:5054/metrics | wc -l
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 83667  100 83667    0     0  12782      0  0:00:06  0:00:06 --:--:-- 19269
1217
```

**Comment by karalabe** [comment]
Perhaps some of these data points might point to something interesting:

```
# date && curl -v  http://localhost:5054/metrics > /dev/null
Wed Apr 26 12:01:53 Europe 2023

*   Trying 127.0.0.1:5054...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to localhost (127.0.0.1) port 5054 (#0)
> GET /metrics HTTP/1.1
> Host: localhost:5054
> User-Agent: curl/7.81.0
> Accept: */*
> 
  0     0    0     0    0     0      0      0 --:--:--  0:00:06 --:--:--     0* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< content-type: text/plain
< server: Lighthouse/v4.1.0-693886b/x86_64-linux
< content-length: 187919
< date: Wed, 26 Apr 2023 12:01:53 GMT
< 
{ [43878 bytes data]
100  183k  100  183k    0     0  29028      0  0:00:06  0:00:06 --:--:-- 57450
* Connection #0 to host localhost left intact
```

Here the interesting thing is that the request time (`date` command) is 12:01:53 and the response header also contains this time. So maybe it starts writing the response and then chokes somewhere.

---

If I send a HEAD request, the server seems to hang:

```
# date && curl -v  http://localhost:5054/metrics -X HEAD    
Wed Apr 26 12:03:59 Europe 2023
Warning: Setting custom HTTP method to HEAD with -X/--request may not work the 
Warning: way you want. Consider using -I/--head instead.
*   Trying 127.0.0.1:5054...
* Connected to localhost (127.0.0.1) port 5054 (#0)
> HEAD /metrics HTTP/1.1
> Host: localhost:5054
> User-Agent: curl/7.81.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 405 Method Not Allowed
< content-type: text/plain; charset=utf-8
< content-length: 23
< date: Wed, 26 Apr 2023 12:03:59 GMT
< 

^C
```

---

If I send a POST request, I get an error back immediately:

```
# date && curl -v  http://localhost:5054/metrics -X POST    
Wed Apr 26 12:03:46 Europe 2023
*   Trying 127.0.0.1:5054...
* Connected to localhost (127.0.0.1) port 5054 (#0)
> POST /metrics HTTP/1.1
> Host: localhost:5054
> User-Agent: curl/7.81.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 405 Method Not Allowed
< content-type: text/plain; charset=utf-8
< content-length: 23
< date: Wed, 26 Apr 2023 12:03:46 GMT
< 
* Connection #0 to host localhost left intact
HTTP method not allowed
```

---

Seems the issue is not the network handling part, rather potentially the metrics collection locking up somewhere.

**Comment by karalabe** [comment]
Tried downgrading to 4.0.1 and upgrading to the unstable branch. All of them behave the same.

**Comment by karalabe** [comment]
Running a pre-built binary on the host, not inside docker removes the hang. Seems the metrics system plays weird with docker.

**Comment by karalabe** [comment]
Making some progress:

I pulled the pre-built lighthouse binary into the screwy docker container and started a second lighthouse instance. The second instance (the one using the pre-built binary from github) responds instantly to the metrics. The one cotnained by default in the docker container hangs. Seems the build is broken somehow

**Comment by karalabe** [comment]
Alternatively, my dumbed down mock lighthouse doesn't do something which the real production one does and that's where the metrics chokes. Does the metric subsystem try to access some resource live when listing the metrics?

**Comment by karalabe** [comment]
Found the issue, it's not lighthouse. 

I was running via Unraid and when the NVMe datadir was mounted via an Unraid share (that's how I am supposed to) it took an insane perf hit, but when I used a direct mount point, it took the metrics scraping back down to 75ms.

I guess one issue that we might still chalk up to lighthouse is *why* it is trying to access the disk on metrics scraping. IMHO that should be a RAM only operation.
